File: utils/screenshot_helpers.py
from playwright.sync_api import Page

from fixtures.test_data import HN_DYNAMIC_SELECTORS


def hide_hacker_news_dynamic_content(page: Page) -> None:
    page.evaluate(
        """(sels) => {
            sels.forEach(sel => {
                document.querySelectorAll(sel).forEach(el => {
                    el.style.visibility = 'hidden';
                });
            });
        }""",
        HN_DYNAMIC_SELECTORS,
    )

File: utils/screenshot.helpers.ts
import { expect, Page, Locator } from '@playwright/test';

export interface ScreenshotOptions {
  fullPage?: boolean;
  mask?: Locator[];
  maxDiffPixelRatio?: number;
}

export async function hideDynamicElements(page: Page, selectors: string[]) {
  await page.evaluate((sels) => {
    sels.forEach((sel) => {
      document.querySelectorAll(sel).forEach((el) => {
        (el as HTMLElement).style.visibility = 'hidden';
      });
    });
  }, selectors);
}

export function getViewportLabel(width: number): string {
  if (width >= 1024) return 'desktop';
  if (width >= 768) return 'tablet';
  return 'mobile';
}

export async function assertVisualMatch(
  locator: Locator,
  snapshotName: string,
  maxDiffPixelRatio = 0.02,
) {
  await expect(locator).toHaveScreenshot(snapshotName, {
    maxDiffPixelRatio,
    animations: 'disabled',
  });
}
